### ISSUE TITLE
Review -- ALI: Adversarially Learned Inference (GAN).

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#52]
Sik-Ho Tsang. [Review — ALI: Adversarially Learned Inference (GAN)](https://sh-tsang.medium.com/review-ali-adversarially-learned-inference-gan-6e4677667914).

**Comment by NorbertZheng** [comment]
## Overview
Not Only Mapping from Latent Space to Data Space, But Also Mapping from Data Space to Latent Space, Outperforms [DCGAN](https://medium.com/@sh.tsang/review-dcgan-deep-convolutional-generative-adversarial-network-gan-ec390cded63c).

Inthis story, Adversarially Learned Inference, (ALI), by Université de Montréal, Stanford, New York University, and CIFAR Fellow, is briefly reviewed. In this story:
 - The generation network maps samples from stochastic latent variables to the data space.
 - The inference network maps training examples in data space to the space of latent variables.
 - The discriminative network is trained to distinguish between **joint latent/data-space samples** $(z,x)$ from the generative network and joint latent/data-space samples from the inference network.

This is a paper in 2017 ICLR with over 1000 citations.

The idea is the same as [BiGAN](https://medium.com/nerd-for-tech/review-bigan-adversarial-feature-learning-gan-535eb76be2ca), but they are proposed independently and published in the same conference (2017 ICLR). Some papers would cite both ALI and [BiGAN](https://medium.com/nerd-for-tech/review-bigan-adversarial-feature-learning-gan-535eb76be2ca) together when talking about this idea.

**Comment by NorbertZheng** [comment]
## ALI: Overall Structure
![image](https://user-images.githubusercontent.com/37235023/216343010-436c32b9-a81c-4f42-b29f-e49044ccf7b0.png)
The adversarially learned inference (ALI) game.
 - Similar to [BiGAN](https://medium.com/nerd-for-tech/review-bigan-adversarial-feature-learning-gan-535eb76be2ca), in order to match the joint distributions, an adversarial game is played, as shown above.
 - $G_{z}(x)$ is the inference network in ALI, we can treat it as encoder.
 - $G_{x}(z)$ is the generation network in ALI, we can treat it as decoder.

The loss function is as follows:
![image](https://user-images.githubusercontent.com/37235023/216343419-0679d64c-67bc-4576-9c56-d98acb412b52.png)
Joint pairs $(x, z)$ are drawn either from $q(x, z)$ or $p(x, z)$, and a discriminator network learns to discriminate between the two, while the encoder and decoder networks are trained to fool the discriminator.

**Comment by NorbertZheng** [comment]
![image](https://user-images.githubusercontent.com/37235023/216343785-ddf8bb74-e749-4ec3-b569-517247de476b.png)
The bidirectional GAN. If we treat $G_{z}(x)$ and $G_{x}(z)$ in ALI as encoder $E$ and decoder(generator) $G$ respectively, it is a bidirectional GAN ([BiGAN](https://medium.com/nerd-for-tech/review-bigan-adversarial-feature-learning-gan-535eb76be2ca)).

Unlike the [GAN](https://medium.com/@sh.tsang/review-gan-generative-adversarial-nets-gan-e12793e1fb75) where the discriminator sees only $x$ as input, in the [BiGAN](https://medium.com/nerd-for-tech/review-bigan-adversarial-feature-learning-gan-535eb76be2ca)/ALI, D sees both $x$ and $z$ , i.e., the observation and its latent representation together.

For a true sample, $x$ is given (it is taken from the training set) and the corresponding $z$ is generated by the encoder $E$.

For a fake sample, $z$ is given (it is sampled from $p(z)$ ) and its corresponding $x$ is generated by the generator $G$.

 - The encoder $E$ is also implemented as a deep neural network and (as in the [AE](https://sh-tsang.medium.com/review-autoencoder-reducing-the-dimensionality-of-data-with-neural-networks-data-visualization-fc16446ae32d)) its architecture is usually taken as the inverse of $G$.
 - It is trained just like the generator, namely by back-propagating from the loss function defined at the output of the discriminator.
 - $\bigoplus$ is the vector concatenation operation to concatenate the flatten x and latent vector z before input into the discriminator.

Once training is complete, just like we can use the generator to predict x for new z, we can use the encoder to predict z for any x.

**Comment by NorbertZheng** [comment]
Maybe we can use such a ALI training objective during replay process to improve wakesleep replay mode?

**Comment by NorbertZheng** [comment]
### Latent Space Interpolations
![image](https://user-images.githubusercontent.com/37235023/216347594-adee1d03-25d7-45fc-b2a4-bc01105a40c0.png)
Latent space interpolations on the CelebA validation set.

 - By linearly interpolating between $z_{1}$ and $z_{2}$ and passing the intermediary points through the decoder, the above plot is generated at the input-space interpolations.
 - Smooth transitions are observed between pairs of examples, and intermediary images remain believable (pixel-level, not senmantic-level?).

**Comment by NorbertZheng** [comment]
### Semi-Supervised Learning
![image](https://user-images.githubusercontent.com/37235023/216348082-6dea0517-79fb-460d-b649-54ad653bb9b6.png)
SVHN test set missclassification rate.

 - Using ALI’s inference network to extract features, then using SVM to predict the class using ALI’s extracted features, a misclassification rate is achieved that is roughly 3.00% lower than [DCGAN](https://medium.com/@sh.tsang/review-dcgan-deep-convolutional-generative-adversarial-network-gan-ec390cded63c).

![image](https://user-images.githubusercontent.com/37235023/216348598-a7090d05-3cd9-47e6-840a-10c6ec36c379.png)
CIFAR10 test set missclassification rate for semi-supervised learning using different numbers of trained labeled examples.

 - ALI’s performance is investigated as well when label information is taken into account during training.
 - The discriminator takes $x$ and $z$ as input and outputs a distribution over $K+1$ classes, where $K$ is the number of categories.
 - The above table shows that ALI offers a modest improvement over Salimans et al. (2016), more specifically for 1000 and 2000 labeled examples.

It is conjectured that the latent representation learned by ALI is better untangled with respect to the classification task and that it generalizes better.

**Comment by NorbertZheng** [comment]
### Conditional Generation
![image](https://user-images.githubusercontent.com/37235023/216348892-7ed88d7d-465d-4619-87bc-8f10cd1d3685.png)
Conditional generation sequence.
 - ALI is extended to match a conditional distribution where $y$ represent a fully observed conditioning variable, e.g. attributes in CelebA dataset.

The corresponding loss function is as follows:
![image](https://user-images.githubusercontent.com/37235023/216349058-d9a326a3-4df6-4ba9-a9f8-52a6b0a31974.png)

 - We can treat this as ALI+[CGAN](https://medium.com/@sh.tsang/review-cgan-conditional-gan-gan-78dd42eee41).
 - (I) to (IV): A single fixed latent code $z$ is sampled.
 - (a) to (l): Attributes are then varied uniformly over rows across all columns in the following sequence: (b) black hair; (c) brown hair; (d) blond hair; (e) black hair, wavy hair; (f) blond hair, bangs; (g) blond hair, receding hairline; (h) blond hair, balding; (i) black hair, smiling; (j) black hair, smiling, mouth slightly open; (k) black hair, smiling, mouth slightly open, eyeglasses; (l) black hair, smiling, mouth slightly open, eyeglasses, wearing hat.
 - (Since ALI is similar to [BiGAN](https://medium.com/nerd-for-tech/review-bigan-adversarial-feature-learning-gan-535eb76be2ca), I don’t describe much. If interested, please feel free to read the paper for more details.).

**Comment by NorbertZheng** [comment]
## Reference
[2017 ICLR] [ALI] [Adversarially Learned Inference](https://arxiv.org/abs/1606.00704).

**Comment by NorbertZheng** [comment]
## Generative Adversarial Network (GAN)
Image Synthesis [[GAN](https://medium.com/@sh.tsang/review-gan-generative-adversarial-nets-gan-e12793e1fb75)] [[CGAN](https://medium.com/@sh.tsang/review-cgan-conditional-gan-gan-78dd42eee41)] [[LAPGAN](https://medium.com/@sh.tsang/review-lapgan-laplacian-generative-adversarial-network-gan-e87200bbd827)] [[AAE](https://sh-tsang.medium.com/review-aae-adversarial-autoencoders-gan-e8fda9160542)] [[DCGAN](https://medium.com/@sh.tsang/review-dcgan-deep-convolutional-generative-adversarial-network-gan-ec390cded63c)] [[CoGAN](https://sh-tsang.medium.com/review-cogan-coupled-generative-adversarial-networks-gan-273f70b340af)] [[SimGAN](https://sh-tsang.medium.com/review-simgan-learning-from-simulated-and-unsupervised-images-through-adversarial-training-gan-86a7003add50)] [[BiGAN](https://medium.com/nerd-for-tech/review-bigan-adversarial-feature-learning-gan-535eb76be2ca)] [[ALI](https://sh-tsang.medium.com/review-ali-adversarially-learned-inference-gan-6e4677667914)]
Image-to-image Translation [[Pix2Pix](https://sh-tsang.medium.com/review-pix2pix-image-to-image-translation-with-conditional-adversarial-networks-gan-ac85d8ecead2)] [[UNIT](https://sh-tsang.medium.com/review-unit-unsupervised-image-to-image-translation-networks-gan-4a25ced6d078)]
Super Resolution [[SRGAN & SRResNet](https://medium.com/@sh.tsang/review-srgan-srresnet-photo-realistic-super-resolution-gan-super-resolution-96a6fa19490)] [[EnhanceNet](https://medium.com/@sh.tsang/reading-enhancenet-automated-texture-synthesis-super-resolution-8429635aa75e)] [[ESRGAN](https://medium.com/towards-artificial-intelligence/reading-esrgan-enhanced-super-resolution-generative-adversarial-networks-super-resolution-e8533ad006b5)]
Blur Detection [[DMENet](https://sh-tsang.medium.com/review-dmenet-deep-defocus-map-estimation-using-domain-adaptation-blur-detection-20fdcaf5e384)]
Camera Tampering Detection [[Mantini’s VISAPP’19](https://sh-tsang.medium.com/review-mantinis-visapp-19-generative-reference-model-and-deep-learned-features-camera-f608371c9854)]
Video Coding [[VC-LAPGAN](https://medium.com/@sh.tsang/reading-vc-lapgan-video-coding-oriented-laplacian-pyramid-of-generative-adversarial-networks-74daa2d23d3c)] [[Zhu TMM’20](https://sh-tsang.medium.com/review-zhu-tmm20-generative-adversarial-network-based-intra-prediction-for-video-coding-c8a217c564ea)] [[Zhong ELECGJ’21](https://sh-tsang.medium.com/review-zhong-elecgj21-a-gan-based-video-intra-coding-hevc-intra-9e3486dbca78)]
